control/discovery.py

```diff
@@ -0,0 +1,958 @@
+        GLOBAL_CNLID += 1
+        GLOBAL_GEN_CNT += 1
+        lock.release()
+
+        while True:
+            sq_head_ptr += 1
+            if sq_head_ptr > SQ_HEAD_MAX:
+                sq_head_ptr = 1
+            head = sock.recv(NVME_TCP_PDU_UNIT)
+            if not head:
+                time.sleep(0.01)
+                continue
+
+            PDU = struct.unpack_from('<BBBBI', head, 0)
+            pdu_type = PDU[0]
+            PSH_flag = PDU[1]
+            PH_len = PDU[2]
+            PH_off = PDU[3]
+            package_len = PDU[4]
+
+            # if the length exceeds one packet, continue to recive
+            buffer = [head]
+            if package_len > NVME_TCP_PDU_UNIT:
+                i = package_len // NVME_TCP_PDU_UNIT
+                if package_len % NVME_TCP_PDU_UNIT == 0:
+                    i -= 1
+                while i > 0:
+                    d = sock.recv(NVME_TCP_PDU_UNIT)
+                    if d:
+                        buffer.append(d)
+                        i -= 1
+                    else:
+                        break
+            data = b''.join(buffer)
+
+            # ICreq
+            if pdu_type == NVME_TCP_ICREQ:
+                err = self.reply_initialize(sock)
+
+            # CMD
+            if pdu_type == NVME_TCP_CMD:
+                CMD1 = struct.unpack_from('<BBH', data, 8)
+                opcode = CMD1[0]
+                reserved = CMD1[1]
+                cmd_id = CMD1[2]
+
+                # fabric command
+                if opcode == NVME_FABRIC_OPC:
+                    fabric_cmd_type = struct.unpack_from('<B', data, 12)[0]
+
+                    if fabric_cmd_type == NVME_FCTYPE_CONNECT:
+                        (err, nvmeof_connect_data_hostid, nvmeof_connect_data_cntlid,
+                            nvmeof_connect_data_subnqn, nvmeof_connect_data_hostnqn) = \
+                            self.reply_fc_cmd_connect(sock, data, self_cnlid, sq_head_ptr)
+
+                    if fabric_cmd_type == NVME_FCTYPE_PROP_GET:
+                        err, shutdown_now = self.reply_fc_cmd_prop_get(sock, data, sq_head_ptr,
+                                                                       cmd_id, shutdown_notification)
+
+                    if fabric_cmd_type == NVME_FCTYPE_PROP_SET:
+                        err, shutdown_notification = self.reply_fc_cmd_prop_set(sock, data,
+                                                                                sq_head_ptr, cmd_id)
+
+                    if fabric_cmd_type == NVME_FCTYPE_AUTH_SEND:
+                        self.logger.error("can't handle NVME_FCTYPE_AUTH_SEND request.")
+                    if fabric_cmd_type == NVME_FCTYPE_AUTH_RECV:
+                        self.logger.error("can't handle NVME_FCTYPE_AUTH_RECV request.")
+                    if fabric_cmd_type == NVME_FCTYPE_DISCONNECT:
+                        self.logger.error("can't handle NVME_FCTYPE_DISCONNECT request.")
```

Comment:
All these checks should be replaced with if-elif, otherwise it'll be doing unnecessary checks.

```suggestion
                    if fabric_cmd_type == NVME_FCTYPE_CONNECT:
                        (err, nvmeof_connect_data_hostid, nvmeof_connect_data_cntlid,
                            nvmeof_connect_data_subnqn, nvmeof_connect_data_hostnqn) = \
                            self.reply_fc_cmd_connect(sock, data, self_cnlid, sq_head_ptr)

                    elif fabric_cmd_type == NVME_FCTYPE_PROP_GET:
                        err, shutdown_now = self.reply_fc_cmd_prop_get(sock, data, sq_head_ptr,
                                                                       cmd_id, shutdown_notification)

                    elif fabric_cmd_type == NVME_FCTYPE_PROP_SET:
                        err, shutdown_notification = self.reply_fc_cmd_prop_set(sock, data,
                                                                                sq_head_ptr, cmd_id)

                    elif fabric_cmd_type == NVME_FCTYPE_AUTH_SEND:
                        self.logger.error("can't handle NVME_FCTYPE_AUTH_SEND request.")
                    elif fabric_cmd_type == NVME_FCTYPE_AUTH_RECV:
                        self.logger.error("can't handle NVME_FCTYPE_AUTH_RECV request.")
                    elif fabric_cmd_type == NVME_FCTYPE_DISCONNECT:
                        self.logger.error("can't handle NVME_FCTYPE_DISCONNECT request.")
```

Given that Python lacks `switch/case` statement, it could be replaced with a callback approach:
```
handle_opcode = {
  NVME_AQ_OPC.GET_LOG_PAGE: self.reply_get_log_page,
  NVME_AQ_OPC.IDENTIFY: self.reply_identify,
  ...
}

try:
  handle_opcode[opcode](...)
except KeyError as e:
  raise UnknownNVMEOpcode(e)
```

Reply:
Thank you for such a good suggestion. I have temporarily changed it to `elif`. What should I do if some functions have specific return values for further changing to a `callback` approach. Of course, if I change it to the selectors(https://github.com/CongMinYin/ceph-nvmeof/commit/afd02b2d37b03b76c2b739e849eb455e81676598) I mentioned earlier, there will be no such issue, only return `err`.

Reply:
as of python 3.10, the match statement can now be used
https://docs.python.org/3/reference/compound_stmts.html#the-match-statement

Reply:
@kasserater, that's an interesting feature. However, keep in mind that [RHEL 9's default Python interpreter is 3.9](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html/installing_and_using_dynamic_programming_languages/assembly_installing-and-using-python_installing-and-using-dynamic-programming-languages#:~:text=In%20RHEL%209%2C%20Python%203.9%20is%20the%20default%20Python%20implementation.).

Reply:
@epuertat but can we make sure our container picks up python 3.10 or later? python 3.10 was released in Oct'21...

Reply:
also, although 3.9 is the default, that same page mentions 3.11 as supported on RHEL 9.x